File: Sources/PartoutOpenVPN/Cross/Internal/CDataPath.swift
// SPDX-FileCopyrightText: 2025 Davide De Rosa
//
// SPDX-License-Identifier: GPL-3.0

import Foundation
internal import PartoutOpenVPN_C

// TODO: #155, move more logic to C (replay protection, byte-aligned enc/dec zd)

final class CDataPath {
    private let mode: UnsafeMutablePointer<openvpn_dp_mode>

    private let encBuffer: UnsafeMutablePointer<pp_zd>

    private let decBuffer: UnsafeMutablePointer<pp_zd>

    private let replay: UnsafeMutablePointer<openvpn_replay>

    private let resizeStep: Int

    private let maxPacketId: UInt32

    private var outPacketId: UInt32

    init(mode: UnsafeMutablePointer<openvpn_dp_mode>, peerId: UInt32) {
        self.mode = mode
        openvpn_dp_mode_set_peer_id(mode, peerId)

        let oneKilo = 1024
        encBuffer = pp_zd_create(64 * oneKilo)
        decBuffer = pp_zd_create(64 * oneKilo)
        replay = openvpn_replay_create()
        resizeStep = 1024
        maxPacketId = .max - 10 * UInt32(oneKilo)
        outPacketId = .zero
    }

    deinit {
        openvpn_replay_free(replay)
        openvpn_dp_mode_free(mode)
        pp_zd_free(encBuffer)
        pp_zd_free(decBuffer)
    }
}

private extension CDataPath {
    func configure(
        cipherKey: Data?,
        hmacKey: Data?,
        block: (UnsafePointer<pp_zd>?, UnsafePointer<pp_zd>?) -> Void
    ) {
        let ck = cipherKey.map { data in
            data.withUnsafeBytes { ptr in
                pp_zd_create_from_data(ptr.bytePointer, data.count)
            }
        }
        let hk = hmacKey.map { data in
            data.withUnsafeBytes { ptr in
                pp_zd_create_from_data(ptr.bytePointer, data.count)
            }
        }
        block(ck, hk)
        if let ck {
            pp_zd_free(ck)
        }
        if let hk {
            pp_zd_free(hk)
        }
    }
}

// MARK: - Bulk encrypt/decrypt

extension CDataPath {
    func encrypt(_ packets: [Data], key: UInt8) throws -> [Data] {
        try packets.map {
            outPacketId += 1
            return try assembleAndEncrypt(
                $0,
                key: key,
                packetId: outPacketId,
                buf: nil
            )
        }
    }

    func decrypt(_ packets: [Data]) throws -> (packets: [Data], keepAlive: Bool) {
        var keepAlive = false
        let list = try packets.compactMap { encrypted -> Data? in

            // framing will throw if compressed (handled in openvpn_dp_framing_parse_*)
            let tuple = try decryptAndParse(
                encrypted,
                buf: nil
            )

            // throw on packet id overflow
            guard tuple.packetId <= maxPacketId else {
                throw OpenVPNDataPathError.overflow
            }
            // ignore replayed packet ids
            guard !openvpn_replay_is_replayed(replay, tuple.packetId) else {
                return nil
            }
            // detect keep-alive packet (ping)
            if tuple.isKeepAlive {
                keepAlive = true
                return nil
            }
            return tuple.data
        }
        return (list, keepAlive)
    }
}

// MARK: - Creating buffers (for testing)

extension CDataPath {
    func assembleAndEncrypt(
        _ packet: Data,
        key: UInt8,
        packetId: UInt32,
        withNewBuffer: Bool
    ) throws -> Data {
        let buf = withNewBuffer ? pp_zd_create(0) : nil
        return try assembleAndEncrypt(packet, key: key, packetId: packetId, buf: buf)
    }

    func decryptAndParse(
        _ packet: Data,
        withNewBuffer: Bool
    ) throws -> DataPathDecryptedAndParsedTuple {
        let buf = withNewBuffer ? pp_zd_create(0) : nil
        return try decryptAndParse(packet, buf: buf)
    }
}

extension CDataPath {
    func assemble(
        packetId: UInt32,
        payload: Data
    ) -> Data {
        let buf = pp_zd_create(openvpn_dp_mode_assemble_capacity(mode, payload.count))
        defer {
            pp_zd_free(buf)
        }
        return assemble(packetId: packetId, payload: payload, buf: buf)
    }

    func encrypt(
        key: UInt8,
        packetId: UInt32,
        assembled: Data
    ) throws -> Data {
        let buf = pp_zd_create(openvpn_dp_mode_encrypt_capacity(mode, assembled.count))
        defer {
            pp_zd_free(buf)
        }
        return try encrypt(key: key, packetId: packetId, assembled: assembled, buf: buf)
    }

    func decrypt(
        packet: Data
    ) throws -> DataPathDecryptedTuple {
        let buf = pp_zd_create(packet.count)
        defer {
            pp_zd_free(buf)
        }
        return try decrypt(packet: packet, buf: buf)
    }

    func parse(
        decrypted: Data,
        header: inout UInt8,
    ) throws -> Data {
        let buf = pp_zd_create(decrypted.count)
        defer {
            pp_zd_free(buf)
        }
        return try parse(decrypted: decrypted, header: &header, buf: buf)
    }
}

// MARK: - Reusing buffers

extension CDataPath {
    func assembleAndEncrypt(
        _ packet: Data,
        key: UInt8,
        packetId: UInt32,
        buf: UnsafeMutablePointer<pp_zd>?
    ) throws -> Data {
        let buf = buf ?? encBuffer
        resize(buf, for: openvpn_dp_mode_assemble_and_encrypt_capacity(mode, packet.count))
        return try packet.withUnsafeBytes { src in
            var error = openvpn_dp_error()
            let zd = openvpn_dp_mode_assemble_and_encrypt(
                mode,
                key,
                packetId,
                buf,
                src.bytePointer,
                packet.count,
                &error
            )
            guard let zd else {
                throw CDataPathError.error(for: error)
            }
            return Data(zeroing: zd)
        }
    }

    func decryptAndParse(
        _ packet: Data,
        buf: UnsafeMutablePointer<pp_zd>?
    ) throws -> DataPathDecryptedAndParsedTuple {
        let buf = buf ?? decBuffer
        resize(buf, for: packet.count)
        return try packet.withUnsafeBytes { src in
            var packetId: UInt32 = .zero
            var header: UInt8 = .zero
            var keepAlive: Bool = false
            var error = openvpn_dp_error()
            let zd = openvpn_dp_mode_decrypt_and_parse(
                mode,
                buf,
                &packetId,
                &header,
                &keepAlive,
                src.bytePointer,
                packet.count,
                &error
            )
            guard let zd else {
                throw CDataPathError.error(for: error)
            }
            let data = Data(zeroing: zd)
            return DataPathDecryptedAndParsedTuple(packetId, header, keepAlive, data)
        }
    }
}

extension CDataPath {
    func assemble(
        packetId: UInt32,
        payload: Data,
        buf: UnsafeMutablePointer<pp_zd>?
    ) -> Data {
        let buf = buf ?? encBuffer
        let inputCount = payload.count
        resize(buf, for: openvpn_dp_mode_assemble_capacity(mode, inputCount))
        return payload.withUnsafeBytes { input in
            let outLength = openvpn_dp_mode_assemble(
                mode,
                packetId,
                buf,
                input.bytePointer,
                inputCount
            )
            return Data(bytes: buf.pointee.bytes, count: outLength)
        }
    }

    func encrypt(
        key: UInt8,
        packetId: UInt32,
        assembled: Data,
        buf: UnsafeMutablePointer<pp_zd>?
    ) throws -> Data {
        let buf = buf ?? encBuffer
        let inputCount = assembled.count
        resize(buf, for: openvpn_dp_mode_encrypt_capacity(mode, inputCount))
        return try assembled.withUnsafeBytes { input in
            var error = openvpn_dp_error()
            let outLength = openvpn_dp_mode_encrypt(
                mode,
                key,
                packetId,
                buf,
                input.bytePointer,
                inputCount,
                &error
            )
            guard outLength > 0 else {
                throw CDataPathError.error(for: error)
            }
            return Data(bytes: buf.pointee.bytes, count: outLength)
        }
    }

    func decrypt(
        packet: Data,
        buf: UnsafeMutablePointer<pp_zd>?
    ) throws -> DataPathDecryptedTuple {
        let buf = buf ?? decBuffer
        let inputCount = packet.count
        resize(buf, for: inputCount)
        return try packet.withUnsafeBytes { input in
            var packetId: UInt32 = 0
            var error = openvpn_dp_error()
            let outLength = openvpn_dp_mode_decrypt(
                mode,
                buf,
                &packetId,
                input.bytePointer,
                inputCount,
                &error
            )
            guard outLength > 0 else {
                throw CDataPathError.error(for: error)
            }
            let data = Data(bytes: buf.pointee.bytes, count: outLength)
            return DataPathDecryptedTuple(packetId, data)
        }
    }

    func parse(
        decrypted: Data,
        header: inout UInt8,
        buf: UnsafeMutablePointer<pp_zd>?
    ) throws -> Data {
        let buf = buf ?? decBuffer
        var inputCopy = [UInt8](decrypted) // copy because parsed in place
        let inputCount = inputCopy.count
        resize(buf, for: inputCount)

        let decryptedOriginal = decrypted
        let parsed = try inputCopy.withUnsafeMutableBytes { input in
            var error = openvpn_dp_error()
            let outLength = openvpn_dp_mode_parse(
                mode,
                buf,
                &header,
                input.bytePointer,
                inputCount,
                &error
            )
            guard outLength > 0 else {
                throw CDataPathError.error(for: error)
            }
            return Data(bytes: buf.pointee.bytes, count: outLength)
        }
        // this should never ever fail because of Swift compile checks
        assert(decryptedOriginal == decrypted, "Parsing is done in-place, work on a copy of decrypted")
        return parsed
    }
}

// MARK: -

private extension CDataPath {
    func resize(_ buf: UnsafeMutablePointer<pp_zd>, for count: Int) {
        guard buf.pointee.length < count else {
            return
        }
        var newCount = count + resizeStep
        newCount -= newCount % resizeStep // align to step boundary
        pp_zd_resize(buf, newCount)
    }
}

File: zig/src/openvpn/internal/c_data_path.zig
// SPDX-FileCopyrightText: 2026 Davide De Rosa
//
// SPDX-License-Identifier: GPL-3.0

const std = @import("std");
const c = @import("c.zig").api;
const errors = @import("errors.zig");
const DataPathDecryptedTuple = @import("data_path_decrypted_tuple.zig").DataPathDecryptedTuple;
const DataPathDecryptedAndParsedTuple = @import("data_path_decrypted_and_parsed_tuple.zig").DataPathDecryptedAndParsedTuple;
const data_path_protocol = @import("data_path_protocol.zig");
const DataPathProtocol = data_path_protocol.DataPathProtocol;
const DataPathDecryptResult = @import("data_path_decrypt_result.zig").DataPathDecryptResult;
const DataPathTestingProtocol = @import("data_path_testing_protocol.zig").DataPathTestingProtocol;

extern fn partout_openvpn_dp_mode_decrypt_and_parse(
    mode: *c.openvpn_dp_mode,
    buffer: *c.pp_zd,
    packet_id: *u32,
    header: *u8,
    keep_alive: *bool,
    source: [*c]const u8,
    source_length: usize,
    native_error: *c.openvpn_dp_error,
) ?*c.pp_zd;

/// C-backed OpenVPN data path.
///
/// `mode` ownership transfers to `create` on success and is ultimately
/// released by `destroy`. Cryptographic/framing transforms, replay-window
/// bookkeeping, and ping recognition delegate to the existing C routines;
/// this type only owns buffers and orchestrates packet batches.
pub const CDataPath = struct {
    allocator: std.mem.Allocator,
    mode: *c.openvpn_dp_mode,
    enc_buffer: *c.pp_zd,
    dec_buffer: *c.pp_zd,
    replay: *c.openvpn_replay,
    out_packet_id: u32 = 0,

    const resize_step: usize = 1024;
    const initial_buffer_size: usize = 64 * 1024;
    const max_packet_id: u32 = std.math.maxInt(u32) - 10 * 1024;

    pub fn create(
        allocator: std.mem.Allocator,
        mode: *c.openvpn_dp_mode,
        peer_id: u32,
    ) std.mem.Allocator.Error!*CDataPath {
        const self = try allocator.create(CDataPath);
        c.openvpn_dp_mode_set_peer_id(mode, peer_id);
        self.* = .{
            .allocator = allocator,
            .mode = mode,
            .enc_buffer = c.pp_zd_create(initial_buffer_size),
            .dec_buffer = c.pp_zd_create(initial_buffer_size),
            .replay = c.openvpn_replay_create(),
        };
        return self;
    }

    pub fn destroy(self: *CDataPath) void {
        const allocator = self.allocator;
        c.openvpn_replay_free(self.replay);
        c.openvpn_dp_mode_free(self.mode);
        c.pp_zd_free(self.enc_buffer);
        c.pp_zd_free(self.dec_buffer);
        allocator.destroy(self);
    }

    pub fn asProtocol(self: *CDataPath) DataPathProtocol {
        return .{ .ptr = self, .vtable = &data_path_vtable };
    }

    pub fn asTestingProtocol(self: *CDataPath) DataPathTestingProtocol {
        return .{
            .data_path = self.asProtocol(),
            .testing_vtable = &testing_vtable,
        };
    }

    pub fn encryptPackets(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        packets: []const []const u8,
        key: u8,
    ) anyerror![][]u8 {
        var result: std.ArrayList([]u8) = .empty;
        errdefer {
            for (result.items) |packet| allocator.free(packet);
            result.deinit(allocator);
        }
        try result.ensureTotalCapacity(allocator, packets.len);
        for (packets) |packet| {
            self.out_packet_id = std.math.add(u32, self.out_packet_id, 1) catch
                return error.DataPathOverflow;
            const encrypted = try self.assembleAndEncrypt(
                allocator,
                packet,
                key,
                self.out_packet_id,
            );
            result.appendAssumeCapacity(encrypted);
        }
        return result.toOwnedSlice(allocator);
    }

    pub fn decryptPackets(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        packets: []const []const u8,
    ) anyerror!DataPathDecryptResult {
        var result: std.ArrayList([]u8) = .empty;
        errdefer {
            for (result.items) |packet| allocator.free(packet);
            result.deinit(allocator);
        }
        try result.ensureTotalCapacity(allocator, packets.len);
        var keep_alive = false;
        for (packets) |packet| {
            var tuple = try self.decryptAndParse(allocator, packet);
            if (tuple.packet_id > max_packet_id) {
                tuple.deinit(allocator);
                return error.DataPathOverflow;
            }
            if (c.openvpn_replay_is_replayed(self.replay, tuple.packet_id)) {
                tuple.deinit(allocator);
                continue;
            }
            if (tuple.is_keep_alive) {
                keep_alive = true;
                tuple.deinit(allocator);
                continue;
            }
            result.appendAssumeCapacity(tuple.data);
            tuple.data = @constCast(&[_]u8{});
        }
        return .{
            .packets = try result.toOwnedSlice(allocator),
            .keep_alive = keep_alive,
        };
    }

    pub fn assemble(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        packet_id: u32,
        payload: []const u8,
    ) anyerror![]u8 {
        const capacity = c.openvpn_dp_mode_assemble_capacity(self.mode, payload.len);
        self.resize(self.enc_buffer, capacity);
        const length = c.openvpn_dp_mode_assemble(
            self.mode,
            packet_id,
            self.enc_buffer,
            payload.ptr,
            payload.len,
        );
        return allocator.dupe(u8, self.enc_buffer.*.bytes[0..length]);
    }

    pub fn encrypt(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        key: u8,
        packet_id: u32,
        assembled: []const u8,
    ) anyerror![]u8 {
        const capacity = c.openvpn_dp_mode_encrypt_capacity(self.mode, assembled.len);
        self.resize(self.enc_buffer, capacity);
        var native_error = emptyNativeError();
        const length = c.openvpn_dp_mode_encrypt(
            self.mode,
            key,
            packet_id,
            self.enc_buffer,
            assembled.ptr,
            assembled.len,
            &native_error,
        );
        if (length == 0) return nativeError(native_error);
        return allocator.dupe(u8, self.enc_buffer.*.bytes[0..length]);
    }

    pub fn assembleAndEncrypt(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        packet: []const u8,
        key: u8,
        packet_id: u32,
    ) anyerror![]u8 {
        const capacity = c.openvpn_dp_mode_assemble_and_encrypt_capacity(self.mode, packet.len);
        self.resize(self.enc_buffer, capacity);
        var native_error = emptyNativeError();
        const data = c.openvpn_dp_mode_assemble_and_encrypt(
            self.mode,
            key,
            packet_id,
            self.enc_buffer,
            packet.ptr,
            packet.len,
            &native_error,
        ) orelse return nativeError(native_error);
        defer c.pp_zd_free(data);
        return allocator.dupe(u8, data.*.bytes[0..data.*.length]);
    }

    pub fn decrypt(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        packet: []const u8,
    ) anyerror!DataPathDecryptedTuple {
        self.resize(self.dec_buffer, packet.len);
        var packet_id: u32 = 0;
        var native_error = emptyNativeError();
        const length = c.openvpn_dp_mode_decrypt(
            self.mode,
            self.dec_buffer,
            &packet_id,
            packet.ptr,
            packet.len,
            &native_error,
        );
        if (length == 0) return nativeError(native_error);
        return .init(
            packet_id,
            try allocator.dupe(u8, self.dec_buffer.*.bytes[0..length]),
        );
    }

    pub fn parse(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        decrypted: []const u8,
        header: *u8,
    ) anyerror![]u8 {
        const input = try allocator.dupe(u8, decrypted);
        defer allocator.free(input);
        self.resize(self.dec_buffer, input.len);
        var native_error = emptyNativeError();
        const length = c.openvpn_dp_mode_parse(
            self.mode,
            self.dec_buffer,
            header,
            input.ptr,
            input.len,
            &native_error,
        );
        if (length == 0) return nativeError(native_error);
        return allocator.dupe(u8, self.dec_buffer.*.bytes[0..length]);
    }

    pub fn decryptAndParse(
        self: *CDataPath,
        allocator: std.mem.Allocator,
        packet: []const u8,
    ) anyerror!DataPathDecryptedAndParsedTuple {
        self.resize(self.dec_buffer, packet.len);
        var packet_id: u32 = 0;
        var header: u8 = 0;
        var keep_alive = false;
        var native_error = emptyNativeError();
        const data = partout_openvpn_dp_mode_decrypt_and_parse(
            self.mode,
            self.dec_buffer,
            &packet_id,
            &header,
            &keep_alive,
            packet.ptr,
            packet.len,
            &native_error,
        ) orelse return nativeError(native_error);
        defer c.pp_zd_free(data);
        return .init(
            packet_id,
            header,
            keep_alive,
            try allocator.dupe(u8, data.*.bytes[0..data.*.length]),
        );
    }

    fn resize(_: *CDataPath, buffer: *c.pp_zd, count: usize) void {
        if (buffer.*.length >= count) return;
        const new_count = std.mem.alignForward(usize, count, resize_step);
        c.pp_zd_resize(buffer, new_count);
    }

    fn emptyNativeError() c.openvpn_dp_error {
        return .{
            .dp_code = c.OpenVPNDataPathErrorNone,
            .crypto_code = c.PPCryptoErrorNone,
        };
    }

    fn nativeError(native: c.openvpn_dp_error) anyerror {
        if (native.dp_code == c.OpenVPNDataPathErrorNone) return error.DataPathFailure;
        return errors.CDataPathError.fromNative(native).toError();
    }

    const data_path_vtable = DataPathProtocol.VTable{
        .encrypt = protocolEncrypt,
        .decrypt = protocolDecrypt,
        .deinit = protocolDeinit,
    };

    const testing_vtable = DataPathTestingProtocol.TestingVTable{
        .assemble = testingAssemble,
        .encrypt = testingEncrypt,
        .assemble_and_encrypt = testingAssembleAndEncrypt,
        .decrypt = testingDecrypt,
        .parse = testingParse,
        .decrypt_and_parse = testingDecryptAndParse,
    };

    fn cast(pointer: *anyopaque) *CDataPath {
        return @ptrCast(@alignCast(pointer));
    }

    fn protocolEncrypt(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        packets: []const []const u8,
        key: u8,
    ) anyerror![][]u8 {
        return cast(pointer).encryptPackets(allocator, packets, key);
    }

    fn protocolDecrypt(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        packets: []const []const u8,
    ) anyerror!DataPathDecryptResult {
        return cast(pointer).decryptPackets(allocator, packets);
    }

    fn protocolDeinit(pointer: *anyopaque) void {
        cast(pointer).destroy();
    }

    fn testingAssemble(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        packet_id: u32,
        payload: []const u8,
    ) anyerror![]u8 {
        return cast(pointer).assemble(allocator, packet_id, payload);
    }

    fn testingEncrypt(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        key: u8,
        packet_id: u32,
        assembled: []const u8,
    ) anyerror![]u8 {
        return cast(pointer).encrypt(allocator, key, packet_id, assembled);
    }

    fn testingAssembleAndEncrypt(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        packet: []const u8,
        key: u8,
        packet_id: u32,
    ) anyerror![]u8 {
        return cast(pointer).assembleAndEncrypt(allocator, packet, key, packet_id);
    }

    fn testingDecrypt(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        packet: []const u8,
    ) anyerror!DataPathDecryptedTuple {
        return cast(pointer).decrypt(allocator, packet);
    }

    fn testingParse(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        decrypted: []const u8,
        header: *u8,
    ) anyerror![]u8 {
        return cast(pointer).parse(allocator, decrypted, header);
    }

    fn testingDecryptAndParse(
        pointer: *anyopaque,
        allocator: std.mem.Allocator,
        packet: []const u8,
    ) anyerror!DataPathDecryptedAndParsedTuple {
        return cast(pointer).decryptAndParse(allocator, packet);
    }
};

test "CDataPath mock round-trips individual, compound, and bulk packets" {
    const allocator = std.testing.allocator;
    const peer_id: u32 = 0x01;
    const key: u8 = 0x02;
    const packet_id: u32 = 0x1020;
    const payload = [_]u8{ 0x11, 0x22, 0x33, 0x44 };

    const mode = c.openvpn_dp_mode_ad_create_mock(c.OpenVPNCompressionFramingDisabled);
    const data_path = try CDataPath.create(allocator, mode, peer_id);
    defer data_path.destroy();

    const assembled = try data_path.assemble(allocator, packet_id, &payload);
    defer allocator.free(assembled);
    const encrypted = try data_path.encrypt(allocator, key, packet_id, assembled);
    defer allocator.free(encrypted);
    var decrypted = try data_path.decrypt(allocator, encrypted);
    defer decrypted.deinit(allocator);
    try std.testing.expectEqual(packet_id, decrypted.packet_id);
    try std.testing.expectEqualSlices(u8, assembled, decrypted.data);
    var header: u8 = 0;
    const parsed = try data_path.parse(allocator, decrypted.data, &header);
    defer allocator.free(parsed);
    try std.testing.expectEqualSlices(u8, &payload, parsed);

    const compound = try data_path.assembleAndEncrypt(
        allocator,
        &payload,
        key,
        packet_id,
    );
    defer allocator.free(compound);
    var compound_result = try data_path.decryptAndParse(allocator, compound);
    defer compound_result.deinit(allocator);
    try std.testing.expectEqual(packet_id, compound_result.packet_id);
    try std.testing.expectEqualSlices(u8, &payload, compound_result.data);

    const packets = [_][]const u8{&payload};
    const encrypted_packets = try data_path.encryptPackets(allocator, &packets, key);
    defer data_path_protocol.freePackets(allocator, encrypted_packets);
    var decrypted_packets = try data_path.decryptPackets(allocator, encrypted_packets);
    defer decrypted_packets.deinit(allocator);
    try std.testing.expect(!decrypted_packets.keep_alive);
    try std.testing.expectEqual(@as(usize, 1), decrypted_packets.packets.len);
    try std.testing.expectEqualSlices(u8, &payload, decrypted_packets.packets[0]);
}
